File: src-tauri/src/infrastructure/turso/provisioning.rs
//! Turso Platform API client.
//!
//! Creates and manages databases programmatically via the Turso Platform API.
//! Used by the registration flow to provision per-user databases.

use reqwest::Client;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Errors that can occur during Turso provisioning operations.
#[derive(Debug)]
pub enum ProvisioningError {
    /// HTTP/transport error
    Http(String),
    /// Turso API rate limit hit
    RateLimit,
    /// Database name conflict (retry with new slug)
    Conflict(String),
    /// Authentication/authorization failure
    Auth(String),
}

impl std::fmt::Display for ProvisioningError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ProvisioningError::Http(msg) => write!(f, "HTTP error: {}", msg),
            ProvisioningError::RateLimit => write!(f, "Turso API rate limit exceeded"),
            ProvisioningError::Conflict(msg) => write!(f, "Conflict: {}", msg),
            ProvisioningError::Auth(msg) => write!(f, "Auth error: {}", msg),
        }
    }
}

impl std::error::Error for ProvisioningError {}

/// Information about a Turso database returned by the Platform API.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DatabaseInfo {
    pub name: String,
    #[serde(rename = "hostname")]
    pub hostname: String,
}

/// Generate a unique database slug from an academy name.
///
/// Format: `academy-{normalized-name}-{4-char-suffix}`
///
/// Rules:
/// - Lowercase
/// - Replace spaces and underscores with hyphens
/// - Remove special characters
/// - Limit to 30 characters
/// - Append 4-char random suffix for uniqueness
pub fn generate_db_slug(academy_name: &str) -> String {
    let normalized: String = academy_name
        .to_lowercase()
        .chars()
        .map(|c| match c {
            'a'..='z' | '0'..='9' | '-' => c,
            ' ' | '_' => '-',
            _ => '-',
        })
        .collect();

    // Remove consecutive hyphens
    let clean: String = normalized.chars().fold(String::new(), |mut acc, c| {
        if c == '-' && acc.ends_with('-') {
            // skip duplicate
        } else {
            acc.push(c);
        }
        acc
    });

    // Trim leading/trailing hyphens and limit to 30 chars
    let trimmed = clean
        .trim_matches('-')
        .chars()
        .take(30)
        .collect::<String>();

    // Add 4-char random suffix for uniqueness
    let suffix: String = Uuid::new_v4().to_string().chars().take(4).collect();

    format!("academy-{}-{}", trimmed, suffix)
}

/// Client for the Turso Platform API.
///
/// Used to create databases, generate auth tokens, and manage
/// database lifecycle programmatically.
pub struct TursoProvisioningService {
    api_token: String,
    org: String,
    client: Client,
}

impl TursoProvisioningService {
    /// Create a new provisioning service.
    ///
    /// * `api_token` — Turso Platform API token (from superadmin env)
    /// * `org` — Turso organization slug (e.g., "academix")
    pub fn new(api_token: String, org: String) -> Self {
        Self {
            api_token,
            org,
            client: Client::new(),
        }
    }

    /// Create a new database in the Turso organization.
    ///
    /// `POST /v1/organizations/{org}/databases`
    ///
    /// If the name conflicts (already exists), retries up to 2 more times
    /// with a new slug suffix.
    pub async fn create_database(&self, name: &str) -> Result<DatabaseInfo, ProvisioningError> {
        let url = format!(
            "https://api.turso.tech/v1/organizations/{}/databases",
            self.org
        );

        #[derive(Serialize)]
        struct CreateRequest {
            name: String,
        }

        // Try to create — if name exists, retry with a new slug
        let mut attempt_name = name.to_string();
        for _ in 0..3 {
            let body = CreateRequest {
                name: attempt_name.clone(),
            };

            let response = self
                .client
                .post(&url)
                .header("Authorization", format!("Bearer {}", self.api_token))
                .json(&body)
                .send()
                .await
                .map_err(|e| ProvisioningError::Http(e.to_string()))?;

            if response.status().is_success() {
                return response
                    .json()
                    .await
                    .map_err(|e| ProvisioningError::Http(e.to_string()));
            }

            if response.status() == reqwest::StatusCode::CONFLICT {
                // Name taken, generate new slug
                let suffix: String =
                    Uuid::new_v4().to_string().chars().take(4).collect();
                attempt_name = format!("{}-{}", &name[..name.len().min(25)], suffix);
                continue;
            }

            if response.status() == reqwest::StatusCode::TOO_MANY_REQUESTS {
                return Err(ProvisioningError::RateLimit);
            }

            let status = response.status();
            let body_text = response.text().await.unwrap_or_default();
            return Err(ProvisioningError::Http(format!(
                "Status {}: {}",
                status, body_text
            )));
        }

        Err(ProvisioningError::Conflict(
            "Could not create database: all name attempts conflicted".into(),
        ))
    }

    /// Generate an auth token for a database.
    ///
    /// `POST /v1/organizations/{org}/databases/{name}/auth/tokens`
    ///
    /// Returns the JWT token string.
    pub async fn create_auth_token(&self, db_name: &str) -> Result<String, ProvisioningError> {
        let url = format!(
            "https://api.turso.tech/v1/organizations/{}/databases/{}/auth/tokens",
            self.org, db_name
        );

        #[derive(Serialize)]
        struct TokenRequest {
            permission: String,
        }

        #[derive(Deserialize)]
        struct TokenResponse {
            #[serde(rename = "jwt")]
            jwt: String,
        }

        let response = self
            .client
            .post(&url)
            .header("Authorization", format!("Bearer {}", self.api_token))
            .json(&TokenRequest {
                permission: "full-access".into(),
            })
            .send()
            .await
            .map_err(|e| ProvisioningError::Http(e.to_string()))?;

        if response.status().is_success() {
            let token_resp: TokenResponse = response
                .json()
                .await
                .map_err(|e| ProvisioningError::Http(e.to_string()))?;
            Ok(token_resp.jwt)
        } else {
            let status = response.status();
            Err(ProvisioningError::Auth(format!(
                "Token creation failed: {}",
                status
            )))
        }
    }

    /// List all databases in the organization (for superadmin).
    ///
    /// `GET /v1/organizations/{org}/databases`
    pub async fn list_databases(&self) -> Result<Vec<DatabaseInfo>, ProvisioningError> {
        let url = format!(
            "https://api.turso.tech/v1/organizations/{}/databases",
            self.org
        );

        let response = self
            .client
            .get(&url)
            .header("Authorization", format!("Bearer {}", self.api_token))
            .send()
            .await
            .map_err(|e| ProvisioningError::Http(e.to_string()))?;

        #[derive(Deserialize)]
        struct ListResponse {
            databases: Vec<DatabaseInfo>,
        }

        let list: ListResponse = response
            .json()
            .await
            .map_err(|e| ProvisioningError::Http(e.to_string()))?;

        Ok(list.databases)
    }

    /// Delete a database (when user deletes account).
    ///
    /// `DELETE /v1/organizations/{org}/databases/{name}`
    pub async fn delete_database(&self, db_name: &str) -> Result<(), ProvisioningError> {
        let url = format!(
            "https://api.turso.tech/v1/organizations/{}/databases/{}",
            self.org, db_name
        );

        let response = self
            .client
            .delete(&url)
            .header("Authorization", format!("Bearer {}", self.api_token))
            .send()
            .await
            .map_err(|e| ProvisioningError::Http(e.to_string()))?;

        if response.status().is_success() || response.status() == reqwest::StatusCode::NOT_FOUND {
            Ok(())
        } else {
            let status = response.status();
            Err(ProvisioningError::Http(format!(
                "Delete failed: {}",
                status
            )))
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_db_slug_lowercases() {
        let slug = generate_db_slug("Music School");
        assert!(slug.starts_with("academy-music-school-"));
        assert_eq!(slug.chars().filter(|&c| c == '-').count(), 3); // academy-{name}-{suffix}
    }

    #[test]
    fn test_generate_db_slug_replaces_spaces() {
        let slug = generate_db_slug("My Academy Name");
        assert!(slug.starts_with("academy-my-academy-name-"));
    }

    #[test]
    fn test_generate_db_slug_removes_special_chars() {
        let slug = generate_db_slug("Hello@World!#2024");
        assert!(slug.starts_with("academy-hello-world-2024-"));
    }

    #[test]
    fn test_generate_db_slug_trims_long_names() {
        let long_name = "A very long academy name that should be truncated significantly";
        let slug = generate_db_slug(long_name);
        // academy- prefix (8) + 30 chars + - (1) + 4 chars suffix = ~43
        assert!(slug.len() <= "academy-".len() + 30 + 1 + 4);
        assert!(slug.starts_with("academy-"));
    }

    #[test]
    fn test_generate_db_slug_handles_underscores() {
        let slug = generate_db_slug("music_school");
        assert!(slug.starts_with("academy-music-school-"));
    }

    #[test]
    fn test_generate_db_slug_unique_suffix() {
        let slug1 = generate_db_slug("Test");
        let slug2 = generate_db_slug("Test");
        // Suffixes should be different (random)
        assert_ne!(slug1, slug2);
    }
}

File: web/src/lib/provisioning.ts
import { createClient, type Client } from '@libsql/client';
import { readdir, readFile } from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

/**
 * Per-user Turso database provisioning for web registration.
 *
 * TypeScript port of the desktop `TursoProvisioningService`
 * (`src-tauri/src/infrastructure/turso/provisioning.rs`) plus the standalone
 * `run_migrations_on_db` (`src-tauri/src/infrastructure/turso/connection_manager.rs`).
 *
 * Flow: env gate → createDatabase (409 retried with fresh suffix) →
 * createAuthToken (bare DB name, design D5) → dbUrl = libsql://{hostname} →
 * runMigrationsOnDb (web/migrations/per-user/001..020). On any failure after
 * database creation, a best-effort DELETE is attempted and registration fails
 * closed (no partial account state, no silent degradation).
 */

export interface ProvisionResult {
  dbName: string;
  dbUrl: string;
  dbToken: string;
  hostname: string;
}

export type ProvisioningErrorCode =
  | 'MISSING_ENV'
  | 'HTTP'
  | 'RATE_LIMIT'
  | 'CONFLICT'
  | 'AUTH'
  | 'MIGRATION';

export class ProvisioningError extends Error {
  readonly code: ProvisioningErrorCode;

  constructor(code: ProvisioningErrorCode, message: string) {
    super(message);
    this.name = 'ProvisioningError';
    this.code = code;
  }
}

export interface TursoEnv {
  apiToken: string;
  org: string;
  group: string;
}

const TURSO_API_BASE = 'https://api.turso.tech/v1';
const MAX_CREATE_ATTEMPTS = 4; // initial attempt + up to 3 retries on 409

/**
 * Generate a unique database slug from an academy name.
 *
 * Format: `academy-{normalized-name}-{4-char-hex}`
 *
 * Rules (mirrors `provisioning.rs::generate_db_slug`):
 * - Lowercase
 * - Replace every non `[a-z0-9-]` character with `-`
 * - Collapse consecutive hyphens
 * - Trim leading/trailing hyphens, cap at 30 chars
 * - Append a random 4-char hex suffix for uniqueness
 */
export function generateDbSlug(academyName: string): string {
  const normalized = academyName
    .toLowerCase()
    .split('')
    .map((c) => (/[a-z0-9-]/.test(c) ? c : '-'))
    .join('');

  // Remove consecutive hyphens
  let clean = '';
  for (const c of normalized) {
    if (c === '-' && clean.endsWith('-')) {
      continue;
    }
    clean += c;
  }

  // Trim leading/trailing hyphens and limit to 30 chars
  const trimmed = clean.replace(/^-+|-+$/g, '').slice(0, 30);

  return `academy-${trimmed}-${randomHex4()}`;
}

/**
 * 4-char hex suffix. Mirrors the desktop behavior of taking the first 4 chars
 * of a v4 UUID (always hex digits).
 */
function randomHex4(): string {
  return globalThis.crypto.randomUUID().slice(0, 4);
}

/**
 * Read the Turso provisioning env vars. Fails closed (design D6): if any of
 * TURSO_API_TOKEN / TURSO_ORG / TURSO_GROUP is missing, provisioning must NOT
 * silently degrade — registration is rejected with an actionable error.
 */
export function getTursoEnv(): TursoEnv {
  const apiToken = import.meta.env.TURSO_API_TOKEN;
  const org = import.meta.env.TURSO_ORG;
  const group = import.meta.env.TURSO_GROUP;

  const missing: string[] = [];
  if (!apiToken) missing.push('TURSO_API_TOKEN');
  if (!org) missing.push('TURSO_ORG');
  if (!group) missing.push('TURSO_GROUP');

  if (missing.length > 0) {
    throw new ProvisioningError(
      'MISSING_ENV',
      `Turso provisioning requires ${missing.join(', ')}; registration fails closed`
    );
  }

  return { apiToken: String(apiToken), org: String(org), group: String(group) };
}

/**
 * Create a database in the Turso organization.
 *
 * `POST /v1/organizations/{org}/databases`
 *
 * On 409 (name conflict) retries up to 3 times, generating a fresh 4-char hex
 * suffix each retry (mirrors `provisioning.rs::create_database`: base name is
 * capped at 25 chars before the new suffix). If all attempts conflict, throws
 * CONFLICT.
 */
export async function createDatabase(
  org: string,
  name: string,
  group: string
): Promise<{ name: string; hostname: string }> {
  const { apiToken } = getTursoEnv();
  const url = `${TURSO_API_BASE}/organizations/${encodeURIComponent(org)}/databases`;

  let attemptName = name;
  for (let attempt = 0; attempt < MAX_CREATE_ATTEMPTS; attempt++) {
    const response = await apiRequest(url, {
      method: 'POST',
      apiToken,
      body: JSON.stringify({ name: attemptName, group }),
    });

    if (response.ok) {
      return parseCreateDatabaseResponse(response);
    }

    if (response.status === 409) {
      if (attempt < MAX_CREATE_ATTEMPTS - 1) {
        attemptName = `${name.slice(0, 25)}-${randomHex4()}`;
        continue;
      }
      throw new ProvisioningError(
        'CONFLICT',
        'Could not create database: all name attempts conflicted'
      );
    }

    throw mapStatusError(response.status, await errorBody(response));
  }

  throw new ProvisioningError(
    'CONFLICT',
    'Could not create database: all name attempts conflicted'
  );
}

/**
 * Create a full-access auth token for a database.
 *
 * `POST /v1/organizations/{org}/databases/{dbName}/auth/tokens`
 *
 * `dbName` MUST be the bare database name, not the `libsql://` URL (design D5 —
 * the desktop `lib.rs` passes the URL, which is a known bug; do not replicate).
 */
export async function createAuthToken(org: string, dbName: string): Promise<string> {
  const { apiToken } = getTursoEnv();
  const url =
    `${TURSO_API_BASE}/organizations/${encodeURIComponent(org)}/databases/` +
    `${encodeURIComponent(dbName)}/auth/tokens`;

  const response = await apiRequest(url, {
    method: 'POST',
    apiToken,
    body: JSON.stringify({ permission: 'full-access' }),
  });

  if (response.ok) {
    try {
      const data = (await response.json()) as { jwt?: string };
      if (typeof data?.jwt === 'string' && data.jwt.length > 0) {
        return data.jwt;
      }
    } catch {
      // fall through to the malformed-response error below
    }
    throw new ProvisioningError('HTTP', 'Malformed createAuthToken response: missing jwt');
  }

  throw mapStatusError(response.status, await errorBody(response));
}

/**
 * Delete a database. A 404 is treated as success (nothing to delete).
 *
 * `DELETE /v1/organizations/{org}/databases/{dbName}`
 */
export async function deleteDatabase(org: string, dbName: string): Promise<void> {
  const { apiToken } = getTursoEnv();
  const url =
    `${TURSO_API_BASE}/organizations/${encodeURIComponent(org)}/databases/` +
    `${encodeURIComponent(dbName)}`;

  const response = await apiRequest(url, { method: 'DELETE', apiToken });

  if (response.ok || response.status === 404) {
    return;
  }

  throw mapStatusError(response.status, await errorBody(response));
}

/**
 * Run the per-user migrations against a database.
 *
 * Mirrors the desktop `run_migrations_on_db`: creates a `_schema_migrations`
 * tracking table, reads `web/migrations/per-user/*.sql` sorted by filename,
 * and per file — skips it if already recorded, otherwise executes the whole
 * file (batched) and records the version. Re-running is a no-op. A failing
 * file aborts with a MIGRATION error.
 */
export async function runMigrationsOnDb(client: Client): Promise<void> {
  await client.execute(`CREATE TABLE IF NOT EXISTS _schema_migrations (
    version TEXT PRIMARY KEY,
    applied_at TEXT NOT NULL DEFAULT (datetime('now'))
  )`);

  const dir = migrationsDir();
  let entries: string[];
  try {
    entries = (await readdir(dir)).filter((f) => f.endsWith('.sql')).sort();
  } catch (err) {
    throw new ProvisioningError(
      'MIGRATION',
      `Failed to read migrations dir: ${errorMessage(err)}`
    );
  }

  for (const version of entries) {
    const applied = await client.execute({
      sql: 'SELECT version FROM _schema_migrations WHERE version = ?',
      args: [version],
    });

    if (applied.rows.length > 0) {
      continue;
    }

    let sql: string;
    try {
      sql = await readFile(path.join(dir, version), 'utf8');
    } catch (err) {
      throw new ProvisioningError('MIGRATION', `Failed to read ${version}: ${errorMessage(err)}`);
    }

    try {
      await client.executeMultiple(sql);
    } catch (err) {
      throw new ProvisioningError('MIGRATION', `Migration ${version} failed: ${errorMessage(err)}`);
    }

    try {
      await client.execute({
        sql: 'INSERT INTO _schema_migrations (version) VALUES (?)',
        args: [version],
      });
    } catch (err) {
      throw new ProvisioningError(
        'MIGRATION',
        `Failed to record migration ${version}: ${errorMessage(err)}`
      );
    }
  }
}

/**
 * Provision a per-user database for a newly registered academy.
 *
 * Orchestrates: env gate → createDatabase → createAuthToken (bare name) →
 * dbUrl = `libsql://{hostname}` → runMigrationsOnDb. On any failure after the
 * database was created, a best-effort DELETE is attempted and the error is
 * rethrown as a ProvisioningError (fail closed — no partial state).
 */
export async function provisionUser(academyName: string): Promise<ProvisionResult> {
  const env = getTursoEnv();
  const slug = generateDbSlug(academyName);

  // createDatabase throws on failure without creating a DB, so no cleanup is
  // needed before this point.
  const { name: dbName, hostname } = await createDatabase(env.org, slug, env.group);

  try {
    const dbToken = await createAuthToken(env.org, dbName);
    const dbUrl = `libsql://${hostname}`;
    const client = createClient({ url: dbUrl, authToken: dbToken });
    try {
      await runMigrationsOnDb(client);
    } finally {
      try {
        client.close();
      } catch {
        // Closing the migration connection is best-effort bookkeeping only.
      }
    }
    return { dbName, dbUrl, dbToken, hostname };
  } catch (err) {
    // The DB exists — try to clean it up, but never let cleanup mask the
    // original error (design R5: best-effort DELETE, fail closed).
    try {
      await deleteDatabase(env.org, dbName);
    } catch {
      // best-effort only
    }
    if (err instanceof ProvisioningError) {
      throw err;
    }
    throw new ProvisioningError(
      'HTTP',
      `Provisioning failed after database creation: ${errorMessage(err)}`
    );
  }
}

/** Directory holding the per-user migration copies. */
function migrationsDir(): string {
  const override = process.env.TURSO_MIGRATIONS_DIR;
  if (override) {
    return override;
  }
  // Dev/tests/typecheck: web/src/lib/provisioning.ts → web/migrations/per-user.
  return fileURLToPath(new URL('../../migrations/per-user/', import.meta.url));
}

async function apiRequest(
  url: string,
  init: { method: 'GET' | 'POST' | 'DELETE'; apiToken: string; body?: string }
): Promise<Response> {
  try {
    return await fetch(url, {
      method: init.method,
      headers: {
        Authorization: `Bearer ${init.apiToken}`,
        'Content-Type': 'application/json',
      },
      body: init.body,
    });
  } catch (err) {
    throw new ProvisioningError('HTTP', `Turso API request failed: ${errorMessage(err)}`);
  }
}

interface CreateDatabaseResponse {
  database?: {
    name?: string;
    hostname?: string;
    // Desktop serde expects capitalized keys; accept both shapes.
    Name?: string;
    Hostname?: string;
  };
}

async function parseCreateDatabaseResponse(
  response: Response
): Promise<{ name: string; hostname: string }> {
  let data: CreateDatabaseResponse;
  try {
    data = (await response.json()) as CreateDatabaseResponse;
  } catch (err) {
    throw new ProvisioningError('HTTP', `Malformed createDatabase response: ${errorMessage(err)}`);
  }
  const db = data?.database;
  const name = db?.name ?? db?.Name;
  const hostname = db?.hostname ?? db?.Hostname;
  if (!name || !hostname) {
    throw new ProvisioningError(
      'HTTP',
      'Malformed createDatabase response: missing name/hostname'
    );
  }
  return { name, hostname };
}

function mapStatusError(status: number, body: string): ProvisioningError {
  if (status === 401 || status === 403) {
    return new ProvisioningError('AUTH', `Turso API auth failed (${status}): ${body}`);
  }
  if (status === 429) {
    return new ProvisioningError('RATE_LIMIT', 'Turso API rate limit exceeded (429)');
  }
  return new ProvisioningError('HTTP', `Turso API HTTP ${status}: ${body}`);
}

async function errorBody(response: Response): Promise<string> {
  try {
    return await response.text();
  } catch {
    return '';
  }
}

function errorMessage(err: unknown): string {
  return err instanceof Error ? err.message : String(err);
}
